substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs

```diff
@@ -705,71 +710,109 @@ where
 	async fn submit_events<EventStream>(
 		&mut self,
 		startup_point: &StartupPoint<Block>,
-		mut stream: EventStream,
+		stream: EventStream,
 		sink: Subscription,
 		rx_stop: oneshot::Receiver<()>,
 	) -> Result<(), SubscriptionManagementError>
 	where
-		EventStream: Stream<Item = NotificationType<Block>> + Unpin,
+		EventStream: Stream<Item = NotificationType<Block>> + Unpin + Send,
 	{
-		let mut stream_item = stream.next();
+		// make the stream abortable
+		let (stream, handle) = futures::stream::abortable(stream);
+		let cancel_handle = handle.clone();
 
-		// The stop event can be triggered by the chainHead logic when the pinned
-		// block guarantee cannot be hold. Or when the client is disconnected.
-		let connection_closed = sink.closed();
-		tokio::pin!(connection_closed);
-		let mut stop_event = futures_util::future::select(rx_stop, connection_closed);
+		// create a oneshot to propagate error messages
+		let (tx_send, mut tx_receive) = oneshot::channel();
 
-		while let Either::Left((Some(event), next_stop_event)) =
-			futures_util::future::select(stream_item, stop_event).await
-		{
-			let events = match event {
+		// TODO: is there a cleaner way to do this?
+		let tx_send = Arc::new(parking_lot::Mutex::new(tx_send));
+		let mut events = move |event| {
+			match event {
 				NotificationType::InitialEvents(events) => Ok(events),
 				NotificationType::NewBlock(notification) =>
 					self.handle_import_blocks(notification, &startup_point),
 				NotificationType::Finalized(notification) =>
 					self.handle_finalized_blocks(notification, &startup_point),
 				NotificationType::MethodResponse(notification) => Ok(vec![notification]),
-			};
-
-			let events = match events {
-				Ok(events) => events,
-				Err(err) => {
-					debug!(
-						target: LOG_TARGET,
-						"[follow][id={:?}] Failed to handle stream notification {:?}",
-						self.sub_id,
-						err
-					);
-					_ = sink.send(&FollowEvent::<String>::Stop).await;
-					return Err(err)
-				},
-			};
+			}
+			.map_err(|err| {
+				debug!(
```

Comment:
move this debug down where you send the error on the channel and remove the map_err here?

Reply:
self is owned by the parent closure, so when i try to move it it causes ownership issues

Reply:
ah I see it's probably the subscription ID that you need never mind then.

you could probably switch to https://docs.rs/futures-util/latest/futures_util/future/trait.TryFutureExt.html#method.inspect_err to save a line of code but doesn't matter to me